draft-bishop-httpbis-http2-additional-certs.md

```diff
@@ -377,47 +372,54 @@ contains the new origin as well, and if so, reuse the connection.
+[I-D.ietf-httpbis-origin-frame], or by pushing content) and advertised support
+for HTTP-layer certificates (see {{setting}}), it MAY send a
+`CERTIFICATE_REQUEST` frame describing the desired certificate. Clients MUST NOT
+send requests for the specified origin on the connection until the requested
+certificate has been provided.
 
 ~~~
 Client                                      Server
    <----------------------- (stream 0) ORIGIN --
    -- (stream 0) CERTIFICATE_REQUEST ---------->
    ...
-   -- (stream N) CERTIFICATE_NEEDED ----------->
    <------------------ (stream 0) CERTIFICATE --
-   <-------------- (stream N) USE_CERTIFICATE --
    -- (stream N) GET /from-new-origin --------->
    <----------------------- (stream N) 200 OK --
-
 ~~~
 {: #ex-http2-server-requested title="Client-Requested Certificate"}
 
-Likewise, on each stream where certificate authentication is required,
-the server sends a `CERTIFICATE_NEEDED` frame, which the client
-answers with a `USE_CERTIFICATE` frame indicating the certificate to
-use. If the request parameters or the responding certificate are not
-already available, they will need to be sent as described in
-{{cert-available}} as part of this exchange.
+A server SHOULD provide certificates for an origin before pushing resources from
+it or supplying content referencing the origin. If a client receives a
+`PUSH_PROMISE` referencing an origin for which it has not yet received the
+server's certificate, the client SHOULD verify the server's possession of an
+appropriate certificate by sending a `CERTIFICATE_NEEDED` frame on the pushed
+stream to inform the server that progress is blocked until the request is
+satisfied.
+
+If the certificate is already available, the server SHOULD immediately
+respond with the appropriate `USE_CERTIFICATE` frame. (If the
+certificate has not already been transmitted, the server will need to
+make the certificate available as described in {{cert-available}} before
+completing the exchange.)
+
+If the server does not have the desired certificate, it MUST respond with an
+empty `USE_CERTIFICATE` frame. In this case, or if the server has not advertised
+support for HTTP-layer certificates, or if the client chooses not to request the
+server's certificate, the client MUST NOT process the pushed content.
+
+### Requiring a client certificate
+
+On each stream where certificate authentication is required, the server sends a
+`CERTIFICATE_NEEDED` frame, which the client answers with a `USE_CERTIFICATE`
+frame indicating the certificate to use. If the request parameters or the
+responding certificate are not already available, they will need to be sent as
+described in {{cert-available}} as part of this exchange.
+
+If the client does not have the desired certificate, it MUST respond with an
+empty `USE_CERTIFICATE` frame. In this case, the server will process the request
+using only the certificate provided in the TLS handshake (if any), likely resulting
```

Comment:
On a TLS connection established using a mutual handshake, we need a way to determine if a client certificate sent as part of the TLS handshake is associated to the HTTP request, or if _no_ certificate is associated. The latter could happen for example when the mutual handshake was negotiated for SNI example.com, and then the client sends a request for example.org using the same connection.

For the purpose, it might be a good idea to reserve Cert-ID=0 as a way to indicate the certificate exchanged in the TLS handshake.